x/oracle/keeper/aggregator/aggregator_calculator.go

```diff
@@ -0,0 +1,168 @@
+package aggregator
+
+import (
+	"math/big"
+
+	"github.com/ExocoreNetwork/exocore/x/oracle/keeper/common"
+	"github.com/ExocoreNetwork/exocore/x/oracle/types"
+)
+
+type confirmedPrice struct {
+	sourceId  int32
+	detId     string
+	price     *big.Int
+	timestamp string
+}
+
+// internal struct
+type priceAndPower struct {
+	price *big.Int
+	power *big.Int
+}
+
+// for a specific DS round, it could have multiple values provided by different validators(should not be true if there's no malicious validator)
+type roundPrices struct { //0 means NS
+	detId     string
+	prices    []*priceAndPower
+	price     *big.Int
+	timestamp string
+	// confirmed bool
+}
+
+// udpate priceAndPower for a specific DSRoundID, if the price exists, increase its power with provided data
+// return confirmed=true, when detect power exceeds the threshold
+func (r *roundPrices) updatePriceAndPower(pw *priceAndPower, totalPower *big.Int) (updated bool, confirmed bool) {
+	if r.price != nil {
+		confirmed = true
+		return
+	}
+	for _, item := range r.prices {
+		if item.price.Cmp(pw.price) == 0 {
+			item.power = new(big.Int).Add(item.power, pw.power)
+			updated = true
+			if common.ExceedsThreshold(item.power, totalPower) {
+				r.price = item.price
+				confirmed = true
+			}
+			return
+		}
+	}
+	if len(r.prices) < cap(r.prices) {
```

Comment:
len(r.prices) < cap(r.prices)
Its a good practice taking  takes into account not exceeding the preset capacity.
Is it necessary to take into account other boundary conditions, such as validation of the input parameters ?